upup/pkg/fi/cloudup/apply_cluster.go

```diff
@@ -1384,5 +1385,67 @@ func (n *nodeUpConfigBuilder) BuildConfig(ig *kops.InstanceGroup, apiserverAddit
 	config.Channels = n.channels
 	config.EtcdManifests = n.etcdManifests[role]
 
+	if ig.Spec.WarmPool != nil || cluster.Spec.WarmPool != nil {
+		config.WarmPoolImages = n.getWarmPoolImages(ig)
+	}
+
 	return config, auxConfig, nil
 }
+
+// getWarmPoolImages returns a list of container images that should be pre-pulled during instance pre-initialization
+func (n *nodeUpConfigBuilder) getWarmPoolImages(ig *kops.InstanceGroup) []string {
+	if ig != nil && ig.Spec.Role != kops.InstanceGroupRoleNode && ig.Spec.Role != kops.InstanceGroupRoleAPIServer {
+		return nil
+	}
+
+	images := map[string]bool{}
+
+	if ig.Spec.WarmPool != nil {
+		for _, image := range ig.Spec.WarmPool.Images {
+			images[image] = true
+		}
+	}
+
+	cluster := n.cluster
+	if cluster != nil && cluster.Spec.WarmPool != nil {
+		for _, image := range cluster.Spec.WarmPool.Images {
+			images[image] = true
+		}
+	}
+
+	// Add component and addon images that impact startup time
+	// TODO: Exclude images that only run on control-plane nodes in a generic way
+	desiredImagePrefixes := []string{
+		"602401143452.dkr.ecr.us-west-2.amazonaws.com/",
+		"docker.io/calico/",
```

Comment:
We need to have a think about docker.io images. Unless people configure cache/rewrite assets or similar, these will hit ratelimts quickly.

It could be an idea to properly implement registry auth for containerd before merging this one so people also have the option of using the creds they may have deployed as docker secret.

Reply:
I would like to merge this. kOps 1.21 is in alpha and there is enough time to get feedback on this.
As this desiredImagePrefixes will be a new config fields soon. I would expect will address some of the concerns here.

Reply:
I think we should remove docker.io and prevent docker.io from being used here for now until a solution has been fixed. If not, this will break my test setup very quickly as all nodes are behind the same NAT IPs.

They can be added back once we have a good solution for this.

Reply:
Latest commit should address this.